lib/x86_cpu/x86_env.ml

```diff
@@ -160,6 +162,61 @@ module R32 = struct
   (* Only 8 YMM/XMM registers in x86 32-bit mode *)
   let ymms = Array.sub ymms 0 8
 
+  (* mapping registers to arch variables *)
+  let of_reg reg =
+    let internal_reg = X86_asm_reg.decode reg in
+    match internal_reg with
+    | None -> None
+    | Some reg ->
+      (match reg with
+       | `AL | `AH | `AX | `EAX -> Some rax
+       | `BL | `BH | `BX | `EBX -> Some rbx
+       | `CL | `CH | `CX | `ECX -> Some rcx
+       | `DL | `DH | `DX | `EDX -> Some rdx
+       | `SIL | `SI | `ESI -> Some rsi
+       | `DIL | `DI | `EDI -> Some rdi
+       | `BPL | `BP | `EBP -> Some rbp
+       | `SPL | `SP | `ESP -> Some rsp
+       | `EIP | `IP -> Some rip
+
+       | `GS -> seg_gs
+       | `FS_BASE -> Some fs_base
+       | `CS -> seg_cs
+       | `GS_BASE -> Some gs_base
+       | `DS -> seg_ds
+       | `ES -> seg_es
+       | `SS -> seg_ss
+       | `FS -> seg_fs
+
+       | `XMM0 | `YMM0 -> Some ymms.(0)
+       | `XMM1 | `YMM1 -> Some ymms.(1)
+       | `XMM2 | `YMM2 -> Some ymms.(2)
+       | `XMM3 | `YMM3 -> Some ymms.(3)
+       | `XMM4 | `YMM4 -> Some ymms.(4)
+       | `XMM5 | `YMM5 -> Some ymms.(5)
+       | `XMM6 | `YMM6 -> Some ymms.(6)
+       | `XMM7 | `YMM7 -> Some ymms.(7)
+
+       (** 64-bit registers are invalid *)
+       | `R8B | `R9B | `R10B | `R11B
+       | `R12B | `R13B | `R14B | `R15B
+       | `R8W | `R9W | `R10W | `R11W
+       | `R12W | `R13W | `R14W | `R15W
+       | `R8D | `R9D | `R10D | `R11D
+       | `R12D | `R13D | `R14D | `R15D
+       | `RAX | `RBX | `RCX | `RDX
+       | `RDI | `RSI | `RBP | `RSP
+       | `R8 | `R9 | `R10 | `R11
+       | `R12 | `R13 | `R14 | `R15
+       | `RIP
+       | `XMM8 | `YMM8 | `XMM9 | `YMM9
+       | `XMM10 | `YMM10 | `XMM11 | `YMM11
+       | `XMM12 | `YMM12 | `XMM13 | `YMM13
+       | `XMM14 | `YMM14 | `XMM15 | `YMM15
+       ->
+         assert false (* incorrect architecture *)
```

Comment:
there is no static guarantee that this function wouldn't be called for a wrong architecture. So we either need to raise `invalid_argument` here and specify this clearly, that this function is defined only for x86, or use `result` as the return type, to indicate that the function is not total.

Reply:
`invalid_argument` sounds good, will switch it